File: sample/src/main/java/com/emarsys/sample/MobileEngageFragment.java
package com.emarsys.sample;

import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.Switch;
import android.widget.TextView;
import android.widget.Toast;

import com.emarsys.Emarsys;
import com.emarsys.core.api.ResponseErrorException;
import com.emarsys.core.api.result.CompletionListener;
import com.google.firebase.iid.FirebaseInstanceId;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

public class MobileEngageFragment extends BaseFragment {
    private static final String TAG = "MobileEngageFragment";

    private EditText contactFieldValue;
    private EditText eventName;
    private EditText eventAttributes;
    private EditText messageId;

    private TextView statusLabel;

    @Override
    public String getName() {
        return "Mobile Engage";
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View root = inflater.inflate(R.layout.fragment_mobile_engage, container, false);

        final CompletionListener completionListener = errorCause -> {
            String message;
            if (errorCause == null) {
                message = "OK";
                Log.i(TAG, message);
            } else {
                Log.e(TAG, errorCause.getMessage(), errorCause);
                StringBuilder sb = new StringBuilder();
                if (errorCause instanceof ResponseErrorException) {
                    ResponseErrorException mee = (ResponseErrorException) errorCause;
                    sb.append(mee.getStatusCode());
                    sb.append(" - ");
                }
                sb.append(errorCause.getMessage());
                message = sb.toString();
            }
            statusLabel.append(message);
        };
        statusLabel = root.findViewById(R.id.mobileEngageStatusLabel);

        contactFieldValue = root.findViewById(R.id.contactFieldValue);
        eventName = root.findViewById(R.id.eventName);
        eventAttributes = root.findViewById(R.id.eventAttributes);
        messageId = root.findViewById(R.id.messageId);

        root.findViewById(R.id.appLoginAnonymous).setOnClickListener(v -> {
            Emarsys.setAnonymousContact(completionListener);
            handleRequestSent("Anonymous login: ");
        });

        root.findViewById(R.id.appLogin).setOnClickListener(v -> {
            String id = contactFieldValue.getText().toString();
            Emarsys.setContact(id, errorCause -> {
                completionListener.onCompleted(errorCause);
                if (errorCause != null) {
                    ((MainActivity) getActivity()).updateBadgeCount();
                }
            });
            handleRequestSent("Login: ");
        });

        root.findViewById(R.id.appLogout).setOnClickListener(v -> {
            Emarsys.clearContact(completionListener);
            handleRequestSent("Logout: ");
        });

        root.findViewById(R.id.customEvent).setOnClickListener(v -> {
            String name = eventName.getText().toString();
            String attributesString = eventAttributes.getText().toString();

            Map<String, String> attributes = null;
            if (!attributesString.isEmpty()) {
                try {
                    attributes = new HashMap<>();
                    JSONObject json = new JSONObject(attributesString);
                    Iterator<String> keys = json.keys();
                    while (keys.hasNext()) {
                        String key = keys.next();
                        attributes.put(key, json.getString(key));
                    }
                } catch (JSONException e) {
                    Log.w(TAG, "Event attributes edittext content is not a valid JSON!");
                }
            }

            Emarsys.trackCustomEvent(name, attributes, completionListener);
            handleRequestSent("Custom event: ");
        });

        root.findViewById(R.id.messageOpen).setOnClickListener(v -> {
            Intent intent = new Intent();
            String id = messageId.getText().toString();
            Bundle payload = new Bundle();
            payload.putString("key1", "value1");
            payload.putString("u", String.format("{\"sid\": \"%s\"}", id));
            intent.putExtra("payload", payload);
            Emarsys.Push.trackMessageOpen(intent, completionListener);
            handleRequestSent("Message open: ");
        });

        root.<Switch>findViewById(R.id.doNotDisturb).setOnCheckedChangeListener((buttonView, isChecked) -> {
            if (isChecked) {
                Emarsys.InApp.pause();
            } else {
                Emarsys.InApp.resume();
            }
        });

        root.findViewById(R.id.pushToken).setOnClickListener(view -> FirebaseInstanceId.getInstance().getInstanceId().addOnCompleteListener(
                getActivity(),
                task -> {
                    ClipboardManager clipboard = (ClipboardManager) view.getContext().getSystemService(Context.CLIPBOARD_SERVICE);
                    String pushToken = task.getResult().getToken();
                    String toastMessage = pushToken;
                    if (clipboard != null) {
                        ClipData clip = ClipData.newPlainText("pushtoken", pushToken);
                        clipboard.setPrimaryClip(clip);
                        toastMessage = "Copied: " + toastMessage;
                    }
                    Toast.makeText(
                            view.getContext(),
                            toastMessage,
                            Toast.LENGTH_LONG).show();
                }
        ));

        return root;
    }

    private void handleRequestSent(String title) {
        statusLabel.setText(title);
    }
}

File: new-sample-app/src/main/java/com/emarsys/sample/fragments/MobileEngageFragment.kt
package com.emarsys.sample.fragments

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import com.emarsys.Emarsys
import com.emarsys.sample.R
import com.emarsys.sample.extensions.copyToClipboard
import com.emarsys.sample.extensions.showSnackBar
import com.google.firebase.FirebaseApp
import com.google.firebase.iid.FirebaseInstanceId
import kotlinx.android.synthetic.main.fragment_mobile_engage.*
import org.json.JSONException
import org.json.JSONObject

class MobileEngageFragment : Fragment() {

    private companion object {
        const val TAG = "MobileEngageFragment"
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {

        return inflater.inflate(R.layout.fragment_mobile_engage, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        buttonLogin.setOnClickListener {
            val contactId = contactId.text.toString()
            Emarsys.setContact(contactId) {
                if (it != null) {
                    Log.e(TAG, it.toString())
                    view.showSnackBar("Login: failed :(")
                } else {
                    Log.i(TAG, "Login successful.")
                    loggedInText.append(contactId)
                    loggedInText.visibility = View.VISIBLE
                    view.showSnackBar("Login: OK :)")
                }
            }
        }

        buttonLogout.setOnClickListener {
            Emarsys.clearContact {
                if (it != null) {
                    Log.e(TAG, it.toString())
                } else {
                    val message = "Logout was successful."
                    Log.i(TAG, message)
                    loggedInText.visibility = View.GONE
                    view.showSnackBar(message)
                    loggedInText.text = getString(R.string.logged_in)
                }
            }
        }

        buttonTrackMessage.setOnClickListener {
            val intent = Intent()
            val id = trackMessageId.text.toString()
            val payload = Bundle()
            payload.putString("key1", "value1")
            payload.putString("u", String.format("""{"sid":"%s"}""", id))
            intent.putExtra("payload", payload)
            Emarsys.Push.trackMessageOpen(intent) {
                if (it != null) {
                    Log.e(TAG, it.toString())
                } else {
                    Log.i(TAG, "Message id: $id")
                    view.showSnackBar("Message open: OK")
                }
            }
        }

        buttonTrackCustomEvent.setOnClickListener {
            val eventName = eventName.text.toString()
            val stringAttributes = eventAttributes.text.toString()

            val attributes = mutableMapOf<String, String>()
            if (!stringAttributes.isBlank()) {
                try {
                    val jsonObject = JSONObject(stringAttributes)
                    for (key in jsonObject.keys()) {
                        attributes[key] = jsonObject.getString(key)
                    }
                    Emarsys.trackCustomEvent(eventName, attributes) {
                        if (it != null) {
                            Log.e(TAG, it.toString())
                        } else {
                            Log.i(TAG, "Custom Event Track successful")
                            view.showSnackBar("Custom Event Track successful")
                        }
                    }
                } catch (e: JSONException) {
                    view.showSnackBar("Invalid JSON format")
                    e.printStackTrace()
                }
            }
        }

        switchDoNotDisturb.setOnCheckedChangeListener { _, isChecked ->
            if (isChecked) {
                Emarsys.InApp.pause()
            } else {
                Emarsys.InApp.resume()
            }
        }

        buttonGetPushToken.setOnClickListener {
            FirebaseApp.initializeApp(view.context)
            FirebaseInstanceId.getInstance().instanceId.addOnCompleteListener { task ->
                run {
                    val pushToken = task.result?.token

                    pushToken?.copyToClipboard(view.context)

                    view.showSnackBar("Push Token copied: $pushToken")
                }
            }
        }
    }
}
